Make some Option, Result methods unstably const

jhpratt opened the pull request:
The following methods are now unstably const:

- Option::transpose
- Option::flatten
- Result::flatten

While some methods for could likely be made `const` in the future, nearly all of them require something to be dropped at compile-time, which isn't currently supported. The functions listed above should have a trivial path to stabilization.

rust-highfive commented:
r? @JoshTriplett

(rust-highfive has picked a reviewer for you, use r? to override)

jhpratt commented:
@rustbot modify labels to +T-libs +A-const-fn

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1900302426) [(plain)](https://github.com/rust-lang/rust/commit/426c61648c0746ff50d5964986bc2768c0a9be68/checks/1900302426/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |
153 |     convert, fmt, hint, mem,
    |     ^^^^^^^
    |
    = note: `-D unused-imports` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `core`

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1900321168) [(plain)](https://github.com/rust-lang/rust/commit/6fdc4c406c16d0ef12031f3478c5aa39f7dc738a/checks/1900321168/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libc v0.2.85
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.39
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
407 |     pub const fn unwrap_or(self, default: T) -> T {
    |                                  ^^^^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
567 |     pub const fn ok_or<E>(self, err: E) -> Result<T, E> {
    |                                 ^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
700 |     pub const fn and<U>(self, optb: Option<U>) -> Option<U> {
    |                         ^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
700 |     pub const fn and<U>(self, optb: Option<U>) -> Option<U> {
    |                               ^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
797 |     pub const fn or(self, optb: Option<T>) -> Option<T> {
    |                           ^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
797 |     pub const fn or(self, optb: Option<T>) -> Option<T> {
    |                     ^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
    |
851 |         match (self, optb) {
    |               ^^^^^^^^^^^^ constant functions cannot evaluate destructors

error[E0493]: destructors cannot be evaluated at compile-time
    |
983 |         match (self, other) {
    |               ^^^^^^^^^^^^^ constant functions cannot evaluate destructors

```

</details>

jhpratt committed:
Make some Option, Result methods unstably const

The following functions are now unstably const:

- Option::transpose
- Option::flatten
- Result::transpose

jyn514 commented on library/core/src/option.rs:
```diff
@@ -1750,7 +1751,11 @@ impl<T> Option<Option<T>> {
     #[inline]
     #[stable(feature = "option_flattening", since = "1.40.0")]
-    pub fn flatten(self) -> Option<T> {
-        self.and_then(convert::identity)
+    #[rustc_const_unstable(feature = "const_option", issue = "67441")]
+    pub const fn flatten(self) -> Option<T> {
+        match self {
+            Some(inner) => inner,
+            None => None,
+        }
```
Does it make sense to also make `and_then` unstably const so you don't have to change the body?

jhpratt replied:
`and_then` takes a function as an argument, which afaik cannot currently be declared `const`. The RFC for const trait impls hasn't even been merged yet, surprisingly.

CDirkx commented:
> The functions listed above should have a trivial path to stabilization.

These functions implicitly depend on `const_precise_live_drops` (https://github.com/rust-lang/rust/issues/76654#issuecomment-691582684), without which the compiler complains that something might be dropped (even though it isn't). These functions thus cannot be stabilized before `const_precise_live_drops` is stabilized.

joshtriplett commented:
r? @ecstatic-morse

JohnTitor commented:
Triage: @ecstatic-morse seems busy recently, r? @RalfJung could you take a look?

RalfJung commented on library/core/src/result.rs:
```diff
@@ -1233,7 +1233,8 @@ impl<T, E> Result<Option<T>, E> {
     #[inline]
     #[stable(feature = "transpose_result", since = "1.33.0")]
-    pub fn transpose(self) -> Option<Result<T, E>> {
+    #[rustc_const_unstable(feature = "const_result2", reason = "newly added", issue = "57563")]
```
#57563 is the meta tracking issue, it should not be used for any specific feature. If there isn't any tracking issue for const `Result` methods yet, either we need a new one or we could extend the `Option` one to also cover `Result`.

jhpratt replied:
I'll update the `Option` ones to #67441, as you noted in the other comment. I don't see a tracking issue for const `Result` functions, so I'll create one in a bit.

CDirkx replied:
There is #67520

jhpratt replied:
That's fine. I'm creating a relatively thorough one listing all blockers for all methods.

jhpratt replied:
Updated to the newly created tracking issue.

RalfJung commented on library/core/src/option.rs:
```diff
@@ -1275,7 +1275,8 @@ impl<T, E> Option<Result<T, E>> {
     #[inline]
     #[stable(feature = "transpose_result", since = "1.33.0")]
-    pub fn transpose(self) -> Result<Option<T>, E> {
+    #[rustc_const_unstable(feature = "const_option", issue = "67441")]
```
When this lands, we need to remember to add these to https://github.com/rust-lang/rust/issues/67441.

In fact it looks like `transpose` is already listed there, but `flatten` is missing.

RalfJung commented:
This LGTM; Cc @rust-lang/wg-const-eval

jhpratt commented:
With regard to the const-hack label, should we really consider this a hack? I just quickly checked on godbolt, and it looks like the new implementation actually optimizes slightly better at lower opt levels. Not a huge deal either way, but I figure we may as well take a minuscule performance gain where we can.

RalfJung commented:
We typically use that label when code gets changed to be const-compatible. I don't feel competent to evaluate whether the old or the new implementation is better, but I can imagine that avoiding closures helps. @oli-obk what do you think?

oli-obk commented:
I actually prefer the new implementation since imo it's more readable, even if the old version is more compact

RalfJung commented on library/core/src/result.rs:
```diff
@@ -1233,7 +1233,8 @@ impl<T, E> Result<Option<T>, E> {
     #[inline]
     #[stable(feature = "transpose_result", since = "1.33.0")]
-    pub fn transpose(self) -> Option<Result<T, E>> {
+    #[rustc_const_unstable(feature = "const_result", reason = "newly added", issue = "82814")]
```
We don't seem to usually have a `reason` for `rustc_const_unstable`, so no reason to make one up here.

jhpratt replied:
Fixed.

RalfJung commented:
> I actually prefer the new implementation since imo it's more readable, even if the old version is more compact

All right, I tend to agree so I removed the label.

RalfJung commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 79c2b75e88bdab5fccd7d63750ef11aff2f8eaba has been approved by `RalfJung`

<!-- @bors r=RalfJung 79c2b75e88bdab5fccd7d63750ef11aff2f8eaba -->
<!-- homu: {"type":"Approved","sha":"79c2b75e88bdab5fccd7d63750ef11aff2f8eaba","approver":"RalfJung"} -->
